coda-prototype/model.ts

```diff
@@ -1810,6 +1810,135 @@ class StorageManager {
     }
 }
 
+class FileIO {
+    /**
+     * Saves the given string to a file. The file is determined by a file selector UI.
+     * @param {Blob} fileContents Data to write to the file.
+     * @param {(downloadId: number) => void} onDownloadStartedHandler Function to run once the download has started
+     *                                                                successfully.
+     */
+    static saveFile(fileContents : Blob, onDownloadStartedHandler ?: (downloadId : number) => void) {
+        let url = window.URL.createObjectURL(fileContents);
+        console.log("Saving file from URL", url);
+
+        chrome.downloads.download({
+            url: url,
+            saveAs: true
+        }, onDownloadStartedHandler)
+    }
+
+    /**
+     * Exports the given dataset to file on disk.
+     * @param {Dataset} dataset Dataset to save to disk.
+     */
+    static saveDataset(dataset : Dataset) {
+        let eventJSON = {"data": [], "fields":
+            ["id", "timestamp", "owner", "data", "schemeId", "schemeName", "deco_codeValue", "deco_codeId",
+                "deco_confidence", "deco_manual", "deco_timestamp", "deco_author"]
+        }; // TODO: why are rows being referred to as 'events'?
+
+        // For each 'event', add a row to the output for each scheme if schemes exist, or a single row if not.
+        // TODO: Write this in a less-yucky way such that pushing many empty strings is not required
+        for (let event of dataset.events.values()) {
+            if (Object.keys(dataset.schemes).length === 0) { // If there are no schemes:
+                let newEventData = [];
```

Comment:
This code seems super fragile to changes. As a first step, could we put checks in before persisting that the file that we're about to persist has the right shape. That way if[when] we introduce a bug that causes the data to serialise incorectly we won't disrupt peoples' work.

Second step - can we add a test that does a create data, save data, load data - check that we have the same data we started with?

Reply:
Tests seem like a good idea - I manually checked the refactoring worked with one dataset (or at least, as far as is practical manually). We didn't decide on a testing framework.